inheritance.rst: Illustrate the case of abstract derived contract not providing base constructor arguments (#12872)
ethereum/solidity: docs/contracts/inheritance.rst

Specifying arguments in both places is an error.

If a derived contract does not specify the arguments to all of its base
contracts' constructors, it must be declared abstract. In that case, when

> a contract derives from it, its declaration or its constructor must

Reviewer: A declaration would be the whole contract. The text above already refers to this as "inheritance list" so we should stick to that term here:
```suggestion
a contract derives from it, its inheritance list or its constructor must
```

> a contract derives from it, its declaration or its constructor must

Reviewer: This is a bit hard to parse because "contract" and "it" are used multiple times referring to different things. How about this?
```suggestion
another contract derives from it, that contract's declaration or constructor must
```

> provide the necessary parameters for all base classes that haven't had

Reviewer: Or it could be abstract too.

their parameters specified. For example:

.. code-block:: solidity

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity >=0.7.0 <0.9.0;

    contract Base {
        uint x;
        constructor(uint x_) { x = x_; }
    }

    // DerivedAbstract does not specify Base parameters
    abstract contract DerivedAbstract is Base {
        constructor() {}
    }

    // Concrete must specify Base parameters
    contract Concrete is DerivedAbstract {
        constructor(uint y) Base(y * y) {}

>     }

Reviewer: I'm not sure if this example provides enough value here for how much space it takes. It's nearly identical to the one above. Maybe we could just merge the two examples by making `Derived2` abstract and introducing `Derived3`?

Author: Agreed.

So, I just need a little hand holding on how to deal with PRs. I didn't immediately find a good guide, but if you know of one, please point me to it.  Basically, I accepted your very first suggestion and now your second suggestion is "stale" (marked [outdated]) and I can't accept (i.e. [commit suggestion]).

As far as dealing with the PR, I'm not clear on the cli process for that. I have a shallow clone of ethereium/solidity and I did a `gh pr checkout 12872` ... lets say I make a whole bunch of edits ... I can't push to the original PR because I don't have the rights in its repo. What do I do?

Thanks in advance (or for telling me to RTFM somewhere!)

Reviewer: This is really because you're trying to use the github UI pretty intensively here. Normally I only see people doing that for quick edits and for bigger ones it's better to work locally with git. Github's UI clearly has some rough edges for more extensive editing. Unfortunately I do not know a good guide for this workflow and I myself don't use it but if you prefer to do it that way, it's fine. I'll try to suggest some solutions to problems you encounter if I know how.

> Basically, I accepted your very first suggestion and now your second suggestion is "stale" (marked [outdated]) and I can't accept (i.e. [commit suggestion]).

If you applied one suggestion and it modified lines that another suggestion would touch, you have a conflict. I don't think github provides any UI-level way to resolve that but you don't really have to use that apply button. You can change the text in the PR in whatever way that's feasible. E.g. if github lets you change it in an editor, you can just manually copy over my suggestion.

> I can't push to the original PR because I don't have the rights in its repo. What do I do?

That's because you can only push to your fork (i.e. https://github.com/aathan/solidity). That's where your branch is anyway. The code will be added to the main repo (i.e. https://github.com/ethereum/solidity) when the PR is merged. While it's still a PR, its in your repo and you have full control over it.

Author: Yes, I was just surprised the the github cli `gh pr checkout 12872` left me locally checked out in the solidity repo. I guess that's because I ran that command while in a shallow clone of the solidity repo. I'll try running that in a clone of my repo instead, but I think that didn't work. I'll follow up with you on the dev chat instead of polluting this conversation further.
